culane_metric/evaluate.py

```diff
@@ -0,0 +1,205 @@
+        img_data = data_file.readlines()
+    img_data = [line.split() for line in img_data]
+    img_data = [list(map(float, lane)) for lane in img_data]
+    img_data = [[(lane[i], lane[i + 1]) for i in range(0, len(lane), 2)] for lane in img_data]
+    img_data = [lane for lane in img_data if len(lane) >= 2]
+
+    return img_data
+
+
+def load_prediction_list(label_paths, pred_dir):
+    return [load_prediction(os.path.join(pred_dir, path.replace('.json', '.lines.txt'))) for path in label_paths]
+
+
+def load_labels(label_dir):
+    label_paths = helper_scripts.get_files_from_folder(label_dir, '.json')
+    annos = [[add_ys(xs) for xs in spline_creator.get_horizontal_values_for_four_lanes(label_path) if
+              (np.array(xs) >= 0).sum() > 1]  # lanes annotated with a single point are ignored
+             for label_path in label_paths]
+    label_paths = [
+        helper_scripts.get_label_base(p) for p in label_paths
+    ]
+    return np.array(annos, dtype=object), np.array(label_paths, dtype=object)
+
+
+def eval_predictions(pred_dir, anno_dir, width=30, unofficial=True, sequential=False):
+    print('Loading annotation data ({})...'.format(anno_dir))
+    annotations, label_paths = load_labels(anno_dir)
+    print('Loading prediction data ({})...'.format(pred_dir))
+    predictions = load_prediction_list(label_paths, pred_dir)
+    print('Calculating metric {}...'.format('sequentially' if sequential else 'in parallel'))
+    if sequential:
+        results = t_map(partial(culane_metric, width=width, unofficial=unofficial, img_shape=LLAMAS_IMG_RES), predictions,
+                        annotations)
+    else:
+        results = p_map(partial(culane_metric, width=width, unofficial=unofficial, img_shape=LLAMAS_IMG_RES), predictions,
+                        annotations)
+    total_tp = sum(tp for tp, _, _ in results)
+    total_fp = sum(fp for _, fp, _ in results)
+    total_fn = sum(fn for _, _, fn in results)
+    if total_tp == 0:
+        precision = 0
+        recall = 0
+        f1 = 0
+    else:
+        precision = float(total_tp) / (total_tp + total_fp)
+        recall = float(total_tp) / (total_tp + total_fn)
+        f1 = 2 * precision * recall / (precision + recall)
+
+    return {'TP': total_tp, 'FP': total_fp, 'FN': total_fn, 'Precision': precision, 'Recall': recall, 'F1': f1}
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description="Measure CULane's metric on the LLAMAS dataset")
+    parser.add_argument("--pred_dir", help="Path to directory containing the predicted lanes", required=True)
+    parser.add_argument("--anno_dir", help="Path to directory containing the annotated lanes", required=True)
+    parser.add_argument("--width", type=int, default=30, help="Width of the lane")
+    parser.add_argument("--sequential", action='store_true', help="Run sequentially instead of in parallel")
+    parser.add_argument("--unofficial", action='store_true', help="Use a faster but unofficial algorithm")
```

Comment:
How much of a difference is there between the unofficial and original CULanes score? Is it worth it to keep both implementations? In the end, people are mainly going to be interested in comparing approaches

Reply:
I see a few pros and cons for each. The unofficial one should be a bit faster and more accurate but is not exactly how the original metric is implemented (if that's even an issue). The "official" (default one in the script) one is closer to the original implementation. In my experience, my implementation using images (as done in the original implementation) usually has a difference in the order of 0.01% compared to the original one,  while the "unofficial" one has a difference in the order of 1%. However, that 1% difference is not necessarily wrong.

In the end, I think both are fine. Its up to you to choose wheter to remove one or keep both.

Reply:
With the unofficial one, I assume that the predicted values are slightly moved by the interpolation which can lead to slightly different results, right? That may be something that people may complain about.

I'm not too concerned about the additional run-time of the original implementation. Even if it takes a day, I only need to run that once per submission.

Reply:
Yes, but during the development of a model taking a day to compute a metric would be really bad. However, the actual difference here between the two methods is not that large so I think it is fine.